compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -51,41 +51,108 @@ static bool hasIntrinsic(IREE::GPU::TargetAttr target,
 }
 
 static std::optional<IREE::GPU::DataTiledMMAAttr>
-chooseDataTiledMMAAttr(TypeRange elementTypes, IREE::GPU::TargetAttr target) {
-  assert(elementTypes.size() == 3);
+chooseDataTiledMMAAttr(TypeRange eTypes, IREE::GPU::TargetAttr target,
+                       IREE::Encoding::EncodingAttr encoding) {
   using namespace IREE::GPU;
   MLIRContext *ctx = target.getContext();
-  Type lhs = elementTypes[0];
-  Type rhs = elementTypes[1];
-  Type out = elementTypes[2];
-  auto match = [=](MMAIntrinsic intrinsic, int unrollM, int unrollMToSubgroups,
-                   int unrollN, int unrollNToSubgroups,
-                   int unrollK) -> std::optional<DataTiledMMAAttr> {
-    if (!hasIntrinsic(target, intrinsic)) {
-      return std::nullopt;
+
+  //
+  // Step 1: select a MMAIntrinsic.
+  //
+  SmallVector<MMAIntrinsic> candidateIntrinsics = {
```

Comment:
nit: use array since below we don't make any dynamic (re)allocations
```suggestion
  static constexpr MMAIntrinsic candidateIntrinsics[] = {
```